numba/core/datamodel/new_models.py

```diff
@@ -803,6 +807,82 @@ def __init__(self, dmm, fe_type):
         super(NumPyComplexModel, self).__init__(dmm, fe_type, members)
 
 
+@register_default(types.MachineBoolean)
+@register_default(types.MachineBooleanLiteral)
+class MachineBooleanModel(DataModel):
+    _bit_type = ir.IntType(1)
+    _byte_type = ir.IntType(8)
+
+    def get_value_type(self):
+        return self._bit_type
+
+    def get_data_type(self):
+        return self._byte_type
+
+    def get_return_type(self):
+        return self.get_data_type()
+
+    def get_argument_type(self):
+        return self.get_data_type()
+
+    def as_data(self, builder, value):
+        return builder.zext(value, self.get_data_type())
+
+    def as_argument(self, builder, value):
+        return self.as_data(builder, value)
+
+    def as_return(self, builder, value):
+        return self.as_data(builder, value)
+
+    def from_data(self, builder, value):
+        ty = self.get_value_type()
+        resalloca = cgutils.alloca_once(builder, ty)
+        cond = builder.icmp_unsigned('==', value, value.type(0))
+        with builder.if_else(cond) as (then, otherwise):
+            with then:
+                builder.store(ty(0), resalloca)
+            with otherwise:
+                builder.store(ty(1), resalloca)
+        return builder.load(resalloca)
+
+    def from_argument(self, builder, value):
+        return self.from_data(builder, value)
+
+    def from_return(self, builder, value):
+        return self.from_data(builder, value)
+
+@register_default(types.MachineInteger)
+@register_default(types.MachineIntegerLiteral)
+class MachineIntegerModel(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        be_type = ir.IntType(fe_type.bitwidth)
+        super(MachineIntegerModel, self).__init__(dmm, fe_type, be_type)
+
+
+@register_default(types.MachineFloat)
+class MachineFloatModel(PrimitiveModel):
+    def __init__(self, dmm, fe_type):
+        if fe_type.bitwidth <= 32:
+            be_type = ir.FloatType()
+        elif fe_type.bitwidth <= 64:
+            be_type = ir.DoubleType()
+        else:
+            raise NotImplementedError(fe_type)
+        super(MachineFloatModel, self).__init__(dmm, fe_type, be_type)
+
+
+@register_default(types.MachineComplex)
+class MachineComplexModel(ComplexModel):
+    _element_type = NotImplemented
+
+    def __init__(self, dmm, fe_type):
+        members = [
+            ('real', fe_type.underlying_float),
+            ('imag', fe_type.underlying_float),
+        ]
+        super(MachineComplexModel, self).__init__(dmm, fe_type, members)
+
```

Comment:
It's the same across Python/NumPy/Machine, so they can share implementation. You can apply the `@register_default` decorator repeatedly on the same class for different types. Nothing should ever refer to these classes by name.

Reply:
Fixed in b1b8ea8

Reply:
reviewed